### ISSUE TITLE
IP Hopping

### ISSUE THREAD

**Issue by mubix** [scoringengine/scoringengine#403]
It would nice to have the scoreboard be able to take a range, and have it identify configurable number of addresses that it then checks ARP for availability and assumes control over via DHCP (asking for specific addresses) and then some how performs checks randomly from each of the IP addresses

**Comment by mubix** [comment]
I think the best way to do it would be to have it choose one of the IPs for each team per round, keep it on those IPs for the whole round, and then switch once the round is done

**Comment by RustyBower** [comment]
So this would need to be per worker, the engine and web page could still be the same, but we can investigate if it's possible for the workers to hop IP addresses.

**Comment by mubix** [comment]
I don't think it would need to be per-worker. The Scoreboard itself could just set static routes temporarily and flush them each round

**Comment by RustyBower** [comment]
The issue in the current iteration (docker) is that everything is natted through the docker host IP

**Comment by mubix** [comment]
Easy enough to have a script running outside of docker, watching the rounds and picking the IPs and talking to the scoreboard right?

**Comment by RustyBower** [comment]
Yeah. That could definitely work.

**Comment by RustyBower** [comment]
## Feasibility Assessment: IP Hopping

### Current Architecture
The scoring engine runs a single worker container with one IP. All checks execute via `subprocess.run()` and inherit the worker's source IP. The existing `service.worker_queue` field already supports routing checks to specific workers, so the dispatch plumbing is there — the challenge is networking.

### Approach: Multiple Workers with Different IPs

The core idea is the same regardless of platform: run N worker containers, each with a unique IP, and have the engine randomly distribute checks across their Celery queues.

#### Bare Metal / On-Prem

**macvlan network** — gives each container a real IP on the physical network:

```yaml
networks:
  competition:
    driver: macvlan
    driver_opts:
      parent: eth0          # physical NIC on competition network
    ipam:
      config:
        - subnet: 10.0.0.0/24
          ip_range: 10.0.0.200/29   # reserve a small range for workers

services:
  worker-1:
    image: scoringengine/worker
    networks:
      default:          # internal (redis, mysql)
      competition:
        ipv4_address: 10.0.0.200
    environment:
      - SCORINGENGINE_WORKER_QUEUE=pool1

  worker-2:
    networks:
      default:
      competition:
        ipv4_address: 10.0.0.201
    environment:
      - SCORINGENGINE_WORKER_QUEUE=pool2
```

- Each worker gets a real L2 address on the competition VLAN
- Requires host NIC in promiscuous mode
- Won't work inside most VMs or cloud instances

#### AWS / Cloud

macvlan doesn't work in cloud since you don't control L2. Options:

**1. Multiple EC2 instances or ECS tasks (simplest)**
- Run each worker on a separate instance/task
- Each gets its own VPC IP naturally
- Standard horizontal scaling — no networking tricks
- Works with any cloud provider

**2. Multiple ENIs on a single EC2 instance**
- Attach multiple Elastic Network Interfaces, each with a different private IP
- Use source-based routing (`ip rule add from <IP> table <N>`) to bind each worker to a specific ENI
- Limited by instance type (t3.medium = 3 ENIs, c5.xlarge = 4)

**3. Kubernetes with pod-level IPs**
- Each worker pod gets its own IP from the CNI
- Scale with replica count
- Most flexible but requires K8s infrastructure

### Code Changes Needed

Regardless of platform, the application changes are minimal:

1. **Engine task routing** — when dispatching checks, randomly select from a pool of worker queues instead of always using `service.worker_queue`:
   ```python
   queue = random.choice(worker_pool_queues) 
   task = execute_command.apply_async(args=[job], queue=queue)
   ```

2. **Config** — add a `worker_pool_queues` setting (e.g., `pool1,pool2,pool3`)

3. **Docker compose** — template for N worker instances (or a deploy script)

The hard part is the network setup, not the application code. The `worker_queue` routing already exists.

**Comment by ION28** [comment]
Note: An ENI can have many different private IP addresses as well, so the worker node could leverage multiple IPs via a single ENI possibly
